### PR TITLE
Rework Wayland object management

### PR DESCRIPTION
**Pull request by RAOF**
This is a rework of #161, with added goodies along the way:
 * It is now possible to bind to either the highest version available, as long as it's greater than `x`, or to exactly version `x`.
 * A nice compile-time `static_assert` when you forget to add a specialisation for a new Wayland type

### CODE REVIEW

File: include/in_process_server.h [hunk 1]
```diff
@@ -44,6 +44,31 @@ struct zwlr_layer_shell_v1;
...
+ * as wl_output only has a destructor for version >= 3
+ */
+namespace
+{
+void send_release_if_supported(wl_output* to_destroy)
+{
+    if (wl_output_get_version(to_destroy) >= WL_OUTPUT_RELEASE_SINCE_VERSION)
+    {
+        wl_output_release(to_destroy);
+    }
```
[the remark this answers is not in the document]
Comment (RAOF) [reply include/in_process_server.h:62]: Yes, it should.

**Comment by RAOF** [comment]
> I don't fully understand why the template specializations needed to be replaced by a macro, but it works. When you've addressed the nits pull in Alan for final approval.

The template specialisations *haven't* been replaced by a macro? Rather than generating template specialisations for `WlProxy<wl_foo>::WlProxy` with a macro we're instead generating template specialisations for `WlInterfaceDescriptor<wl_foo>` with a macro. That way we can have a fully-generic `Client::bind_if_supported<T>()` (and a nicer `WlProxy<T>`

File: src/xdg_output_v1.cpp [hunk 2]
```diff
@@ -17,18 +17,20 @@
...
+#include "wl_handle.h"
+#include "version_specifier.h"
 #include <boost/throw_exception.hpp>
 
 struct wlcs::XdgOutputManagerV1::Impl
 {
     Impl(Client& client)
         : client{client},
-          manager{client}
+          manager{client.bind_if_supported<zxdg_output_manager_v1>(wlcs::AtLeastVersion{1})}
```
[the remark this answers is not in the document]
Comment (RAOF) [reply src/xdg_output_v1.cpp]: Enjoy `wlcs::AnyVersion`!

File: tests/xdg_surface_stable.cpp [hunk 3]
```diff
@@ -129,10 +129,9 @@ TEST_F(XdgSurfaceStableTest, creating_xdg_surface_from_wl_surface_with_committed
 {
     wlcs::Client client{the_server()};
 
-    auto* const xdg_wm_base = static_cast<struct xdg_wm_base*>(
-        client.acquire_interface("xdg_wm_base", &xdg_wm_base_interface, 1));
+    auto const xdg_wm_base = client.bind_if_supported<struct xdg_wm_base>(AnyVersion);
 
-    auto* const surface = wl_compositor_create_surface(client.compositor());
+    wlcs::WlHandle<wl_surface> const surface{wl_compositor_create_surface(client.compositor())};
```
Comment (AlanGriffiths) [inline tests/xdg_surface_stable.cpp:134]: Not really about this PR, but this looks ugly. Maybe we should have
```c++
auto const surface = wlcs::make_handle(wl_compositor_create_surface(client.compositor()));
```

File: include/wl_interface_descriptor.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ * If an interface requires special handling, a manual specialisation can be
+ * provided (for example, see wl_output handling in in_process_server.h)
+ */
+#define WLCS_CREATE_INTERFACE_DESCRIPTOR(name) \
+    template<> \
+    struct WlInterfaceDescriptor<name> \
+    { \
+        static constexpr bool const has_specialisation = true; \
+        static constexpr wl_interface const* const interface = &name##_interface; \
+        static constexpr void(* const destructor)(name*) = &name##_destroy; \
```
Comment (AlanGriffiths) [inline include/wl_interface_descriptor.h:57]: Using macros for this sort of token pasting is one of the cases metaclasses would fix. At least we're not writing a custom pre-processor. (Yes, TrollTech, I'm looking at you!) Maybe we can replace this in another five years.

File: include/wl_handle.h [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        if (!owns_wl_object)
+        {
+            std::abort();
+        }
+        return proxy;
+    }
+
+    auto wl_proxy() const -> struct wl_proxy*
+    {
+        // This is a precondition failure, but as this is a test-suite lets be generous and make it fail hard and fast
```
Comment (AlanGriffiths) [inline include/wl_handle.h:78]: Pedant says: s/lets/let's/

File: tests/xdg_surface_stable.cpp (hunk 3 above)
Comment (RAOF) [reply tests/xdg_surface_stable.cpp:134]: We already have `wlcs::wrap_wl_object`; I've switched these lines to using it.

**Comment by RAOF** [comment]
bors r=wmww,AlanGriffiths

File: include/wl_handle.h (hunk 5 above)
Comment (RAOF) [reply include/wl_handle.h]: Technically correct. The best kind of correct!

**Comment by RAOF** [comment]
bors r=wmww,AlanGriffiths

ubuntu-devel apparently doesn't know what `groovy-security` is this time?

**Comment by RAOF** [comment]
<William Shatner>Fedora!!!!!!</William Shatner>

bors r=wmww,AlanGriffiths

**Comment by RAOF** [comment]
bors r=wmww,AlanGriffiths

Fedora will *eventually* pass!

**Comment by RAOF** [comment]
Aaaaarrgh

Bors r=wmww,AlanGriffiths

**Comment by RAOF** [comment]
Maybe today?
bors r=wmww,AlanGriffiths

**Pull request merged**